File: 643-maximum-average-subarray-i/maximum-average-subarray-i.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findMaxAverage = function(nums, k) {
    let wAvg = 0;
    let mAvg = Number.NEGATIVE_INFINITY;
    for(let i = 0; i < k; i++) {
        wAvg += nums[i];
    }
    mAvg = Math.max(wAvg/k, mAvg);
    for(let i = k; i < nums.length; i++) {
        wAvg = wAvg - nums[i - k] + nums[i];
        mAvg = Math.max(wAvg/k, mAvg);
    }
    return mAvg;
};

File: 643-maximum-average-subarray-i/maximum-average-subarray-i.java
class Solution {
    public double findMaxAverage(int[] nums, int k) {
        int i = 0;
        int j = 0;
        double max = Integer.MIN_VALUE;
        double sum = 0;
        while(j < nums.length) {
            int len = j - i + 1;
            sum += nums[j];

            while(len > k) {
                sum -= nums[i++];
                len = j - i + 1;
            }
            if(len == k) {
                double avg = (sum/len);
                max = Math.max(max, avg);
            }
            j += 1;
        }
        return max;
    }
}